### PR TITLE
feat(release): compute the version at release time, on demand

### PR DESCRIPTION
**Pull request by marius-cetanas**
Addresses both of Copilot's findings on #23 and reshapes the pipeline so a release consumes **one** version rather than one per push.

## Copilot's findings

**1 — substring grep false-positives.** The preflight grepped `_authToken` as a bare substring, so a `.npmrc` *comment* mentioning it would fail a perfectly good release. Now `scripts/check-npmrc.mjs` matches assignments rather than mentions: the key must start the line or follow a registry prefix, and be followed by `=`; comment lines are skipped. Both directions are tested, including the exact v1.3.0 case (an *empty* `_authToken=` still counts, because an empty token still reads to npm as configured auth).

**2 — the version guard was skippable.** `workflow_dispatch` with `dry_run=false` did a real publish while the consistency check was gated on `github.ref_type == 'tag'`. Since dispatch was about to become the *primary* path, this would have gone from a corner case to the main road. The guard is now unconditional, and there are structural tests asserting it carries no `ref_type` condition and precedes every publish path.

## Versioning moves to release time

Computing per push burns a number for every merge — three merges and one release would jump 1.3.0 → 1.6.0 with 1.4.0 and 1.5.0 never existing. `scripts/next-version.mjs` derives **one** bump from everything since the last release tag (`feat`→minor, `fix`/`perf`→patch, `!`/`BREAKING CHANGE`→major, highest wins), and refuses when nothing is releasable rather than re-publishing the current version.

## Shape

```
Release prepare (manual)  →  derive version, run all checks
                          →  dry_run (default): report and stop
                          →  else: push release/vX.Y.Z branch

you open the PR + merge   ←  the Gated approval

on merge                  →  publish via OIDC → confirm → tag → GitHub release
```

**The tag trigger is gone.** The tag is now an *output* of a successful publish rather than its cause, so a failed publish can no longer leave a tag pointing at a version absent from the registry — exactly what happened to v1.3.0.

**The workflow deliberately does not open the PR.** A PR created with `GITHUB_TOKEN` doesn't trigger workflow runs, so `ci-ok` would never report and branch protection would leave it permanently unmergeable. The job prints a one-click link instead. This is called out in a comment so it doesn't get "fixed" later.

## Verification

330 tests (up from 280), 100% coverage of `src/` maintained, 0 vulnerabilities. 50 of those are new: the resolver, the npmrc guard, and structural assertions over the workflow YAML — so the claims in CLAUDE.md fail the build when they stop being true, per the workspace DoD.

Both scripts were also exercised against real inputs, not just fixtures: the resolver reads actual git history since `v1.3.0` and returns `1.3.1`/patch; the npmrc guard exits 1 on a real assignment and 0 on a comment-only file.

## Note

`scripts/` sits outside the coverage `include` (`src/**`), so it doesn't affect the 100% gate — its tests are still real tests in the same suite.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -32,127 +28,106 @@ jobs:
   publish:
     name: publish to npm
     runs-on: ubuntu-latest
+    # Only a release commit publishes. An ordinary change touching package.json
+    # (a dependency bump, a script) must not reach the registry.
+    if: startsWith(github.event.head_commit.message, 'chore(release):')
```
[the remark this answers is not in the document]
Comment (marius-cetanas) [reply .github/workflows/release.yml]: Good catch, and this one was live rather than theoretical — the repo still allows merge commits, so merging the release PR that way would have made `head_commit` the "Merge pull request …" subject, skipped the job, and published nothing while the documented process looked like it had been followed. A silent no-op is the worst shape for this failure.

Fixed in 831f336: the gate is now `contains(toJSON(github.event.commits.*.message), 'chore(release):')`, so it sees the release commit whether the PR is squashed or merged.

Guarded with three tests in `tests/release/workflow.test.ts` — the condition must not reference `head_commit`, must scan `commits`, and must still key on `chore(release):` rather than degrading into "any push touching package.json".

File: .github/workflows/release-prepare.yml [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+          # Prepend a dated section. Deliberately terse — the PR is where it
+          # gets edited into prose before anyone merges it.
+          node -e '
+            const { readFileSync, writeFileSync } = require("node:fs");
+            const { execFileSync } = require("node:child_process");
+            const next = process.env.NEXT;
+            const date = new Date().toISOString().slice(0, 10);
+            const since = execFileSync("git", ["describe","--tags","--abbrev=0","--match","v*"], {encoding:"utf8"}).trim();
+            const log = execFileSync("git", ["log", `${since}..HEAD`, "--pretty=format:- %s"], {encoding:"utf8"}).trim();
+            const path = "CHANGELOG.md";
```
[the remark this answers is not in the document]
Comment (marius-cetanas) [reply .github/workflows/release-prepare.yml]: Right, and the inconsistency is the telling part: `scripts/next-version.mjs` handles the never-released case deliberately (`lastReleaseTag()` returns `null` and `commitsSince(null)` logs from `HEAD`), but the changelog step then called `git describe --match v*` unguarded and would have thrown before reaching any of that. The support was real but unreachable.

Fixed in 831f336 — the `describe` is wrapped, and the range falls back to `HEAD` when no tag exists.

Verified rather than assumed: ran the exact snippet in a freshly `git init`-ed repo with one commit and no tags, which yields `range: HEAD` and logs the commit instead of throwing.

File: scripts/check-npmrc.mjs [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+// anonymously. The registry answers an anonymous publish with 404 rather than
+// 401 so as not to leak package existence, which makes it look like a missing
+// package rather than an auth failure. See actions/setup-node#1551.
+
+import { existsSync, readFileSync } from "node:fs";
+
+// Matches an assignment, not a mention: the key must be at the start of the
+// line or after a registry prefix (`//registry.npmjs.org/:`), and be followed
+// by `=`. A grep for the bare substring fails a release on a comment.
+const ASSIGNMENT = /(^|[:/])_authtoken\s*=/i;
```
[the remark this answers is not in the document]
Comment (marius-cetanas) [reply scripts/check-npmrc.mjs]: Agreed — the comment directly above the regex claimed start-of-line or a registry-scoped `:`, and the pattern then also accepted `/`, so the code was broader than the rule it documented. That gap is the actual defect here; a value like `cafile=/etc/ssl/_authToken=x` would have failed a release for no reason, and this guard exists to stop false negatives, not to invent false positives.

Narrowed to `(^|:)` in 831f336, which is what the comment said all along. The registry-scoped form still matches, since `//registry.npmjs.org/:_authToken=` puts a `:` immediately before the key.

Two regression tests added: a path value containing `/_authToken=` and a key ending in the word (`legacy_authToken_backup=`) must both be ignored.

**Comment by marius-cetanas** [comment]
Round 2 — three findings, all addressed in aae55d6. Replying to each since they arrived in the review body rather than as inline threads.

**`npm install -g npm@latest` makes the release non-deterministic** — agreed, and this is the one that mattered. A release pipeline that upgrades its own toolchain can change publish behaviour on a run where nothing in the repo changed, which is precisely the property you do not want on the one workflow that does something irreversible. Replaced with an assertion of the 11.5.1 trusted-publishing floor, failing with a message pointing at the Node pin if the runner ever ships something older. Two tests guard it: the workflow must contain no `npm install -g`, and must still assert `11.5.1`.

**Entrypoint detection splits `process.argv[1]` on `/`** (both scripts) — taken, though not for the portability reason. This package is macOS-only and its CI is Linux, so a Windows path was never going to occur. The real defect the finding exposes is that `import.meta.url.endsWith(basename)` is a substring check: it would fire for *any* identically named file, not just this one. Both guards now compare resolved realpaths via `realpathSync` + `fileURLToPath`, which fixes the separator issue as a side effect of fixing the weaker check.

Added `tests/release/cli.test.ts`, which spawns both scripts as subprocesses — importing a module proves nothing about whether its entrypoint block runs.

Separately, and not from review: publishing is now `workflow_dispatch` only. Merging no longer triggers it, so the `head_commit` gate you flagged in round 1 is gone entirely along with the push payload it read.

**Comment by marius-cetanas** [comment]
Round 3 — one finding, fixed in the push above.

**Duplicate annotations from repeated paths** — correct, and it was live rather than hypothetical: `release.yml` passes `${NPM_CONFIG_USERCONFIG:-$HOME/.npmrc}` and `$HOME/.npmrc`, which are the same file on a default runner. A single misconfiguration would have produced two identical `::error` annotations, which reads as two problems. For a guard whose entire value is being believed when it fires, that is worse than cosmetic.

`checkFiles()` now de-duplicates by resolved real path rather than by string, so a symlink collapses onto its target as well. The path in the report stays the one the caller passed, since that is the spelling they have to go and fix.

Seven tests cover it: a repeated path, a symlink pointing at an already-checked file, genuinely distinct offending files still reported separately, and the caller-supplied path surviving into the result.

**Comment by marius-cetanas** [comment]
Round 4 — one finding, and a fair one. Fixed in the push above.

**`release-prepare.yml` still said merging publishes** — in its header comment and, more damagingly, in the job summary. That summary is the instruction a maintainer reads *mid-release*, so following it would have stranded the release: bump merged, nothing shipped, and no indication a step remained. I updated CLAUDE.md and CONTRIBUTING.md when the trigger changed and missed the workflow that describes the process to the person running it.

Both now say merging lands the bump and publishes nothing, and name the `Release` workflow as the next step.

Guarded rather than just corrected, since this is the second time a doc has drifted from the trigger in this PR: a test asserts no workflow or document claims merging publishes, and that `release-prepare.yml` points at `Release`. It earned its place immediately — it failed on first run against `release.yml`, where the comment `Nothing about merging publishes` reads as the opposite of the substring it contains. I reworded the comment rather than loosening the check, because a guard with a carve-out for the phrasing that fooled it is not much of a guard.

**Comment by marius-cetanas** [comment]
Round 5 — one finding, and it does not reproduce. Addressing it anyway, for a different reason.

**Nested double quotes in the `$GITHUB_OUTPUT` lines.** The claim is that the inner `"` terminates the outer string and breaks the step. It does not: command substitution `$(…)` opens a fresh quoting context, so inner double quotes are parsed independently. That is POSIX-specified, not incidental. Verified rather than argued:

```
$ bash -n t.sh   → parses cleanly
$ bash -e t.sh   → exit 0
$ cat out.txt
next=1.4.0
bump=minor
current=1.3.0
```

So the step was never going to fail at runtime, and no release was at risk.

I have rewritten it anyway, as a single `node -e` emitting both lines. Not because it was broken, but because two readers independently read it as broken — and shell that nobody trusts is shell that someone "fixes" incorrectly later. That is a real cost even when the code is right.

Flagging the reasoning explicitly so this does not read as agreement: the original was correct, and the change is for legibility.

---

Separately, CI caught a genuine defect of mine on the previous push, which is worth recording. `tests/release/cli.test.ts` asserted exit 0 from `--force auto` against real git history, but `ci.yml` checks out at `fetch-depth: 1` with no tags — where "nothing releasable" (exit 2) is the *correct* answer. It passed locally against a full clone and failed on all three Node versions. The test was wrong, not the resolver. It now asserts the result is well formed and that the exit code agrees with `releasable`, and I reproduced the CI condition locally with a `--depth 1 --no-tags` clone before and after the fix.

**Comment by marius-cetanas** [comment]
Round 6 — two findings, both correct, both fixed in 2da381b.

**The `_authToken` prefix accepted any `:`.** This is the same gap as the `/` you caught in round 1, and I reintroduced it while fixing that one: the comment said "the `:` of a registry-scoped key" and the pattern said "any colon". Now requires `/:` exactly, so `cafile=C:_authToken=x` and `some:_authToken=x` no longer trip it. Two tests pin the distinction, including that `//registry.npmjs.org/:_authToken=x` must still be caught.

Worth naming the pattern rather than just the bug: this guard has now been over-broad twice, in both cases because the regex drifted from the sentence above it. The tests are written against the sentence, so a third drift fails rather than ships.

**`awk "/^## \[$VERSION\]/"` treats dots as wildcards.** Correct — `1.3.0` also matches `## [1x3x0]`. Low likelihood, but the release body is what a reader trusts to say what changed, so quietly emitting the wrong section is a bad way to fail.

Rather than escaping the dots, I moved extraction into `scripts/release-notes.mjs`, which matches fixed strings. The deciding factor was testability: shell embedded in YAML cannot be tested, which is also why the two earlier findings in these workflows were only caught by reading them. Twelve tests cover it — the wildcard case, a lookalike heading preceding the real one, a longer version sharing a prefix (`1.3.01` vs `1.3.0`), CRLF input, and an empty section — plus a run against the real CHANGELOG.

373 tests, 100% coverage of `src/` held.

**Comment by marius-cetanas** [comment]
Round 7 — one finding, correct, fixed in the push above.

**Every non-zero exit was reported as "No releasable commits".** So a malformed version, missing git history, or any unexpected runtime error would have been diagnosed as "nothing to release" — sending the reader to inspect their commit history rather than the actual fault.

That is the same shape as the failure that cost the v1.3.0 release: an anonymous publish surfaced as a `404`, which reads as "package not found", and I spent a round of this PR looking at the registry instead of the workflow. A wrong error message costs more than no error message, so this is worth more than its size suggests.

The script already exits `2` for "nothing releasable" specifically — the workflow was throwing that information away by using `if !`. It now captures the status and branches: `2` keeps the existing message; anything else non-zero says plainly that it is *not* "nothing to release" and points at the underlying error.

Verified all three paths rather than reasoning about them:

```
releasable      -> exit 0
nothing to ship -> exit 2
malformed input -> exit 1   (previously reported as "nothing to release")
```

Three tests assert the step keys on exit 2 specifically, handles other non-zero exits distinctly, and captures `$?` rather than reverting to `if !`.

**Comment by marius-cetanas** [comment]
Round 8 — two findings, both correct, both fixed in ac8ea53. They are the same defect in two places: a message stating a conclusion the check had not established.

**The tag guard claimed "$PKG is not published."** That was inferred from an `npm view` which also fails on a network or registry error, so a transient outage would have produced a confident, wrong diagnosis. It now reports only what is known — the tag exists, the registry did not report the version — names both possible causes, and says how to tell them apart (`npm view macos-mail-mcp versions`).

**The npmrc guard always blamed `registry-url`.** Also right: the workflow checks `$HOME/.npmrc` and `NPM_CONFIG_USERCONFIG`, so the assignment can just as easily be a user or project `.npmrc` on the runner. Sending someone to edit a workflow that is not the problem wastes the one moment they are paying attention to this guard. The message now names `registry-url` as the cause when the file was generated by the run, and a user or project `.npmrc` otherwise.

The message moved into an exported `describeProblem()` rather than staying inline, so it is testable — five tests cover the file:line, the consequence, both causes, and the instruction.

Worth stating why I am not treating these as nitpicks: this is the third consecutive round on error quality, and that is the right thing to be strict about in this particular PR. The failure that cost the v1.3.0 release was not a missing check — it was a correct refusal wearing a misleading message (`E404`, which reads as "package not found"). Every one of these guards exists to be believed at the exact moment someone is trying to ship, and one that misdirects is worse than one that is silent.

382 tests, 100% coverage of `src/` held.

File: .github/workflows/release-prepare.yml [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+            echo "::error::No releasable commits since the last release. Nothing to publish."
+            cat version.json
+            exit 1
+          fi
+          cat version.json
+          {
+            echo "next=$(node -p "require('./version.json').version")"
+            echo "bump=$(node -p "require('./version.json').bump")"
+            echo "current=$CURRENT"
+          } >> "$GITHUB_OUTPUT"
```
[the remark this answers is not in the document]
Comment (marius-cetanas) [reply .github/workflows/release-prepare.yml]: Checked this rather than taking it at face value, and it does not reproduce: the construct is valid bash. Command substitution `$(…)` opens a fresh quoting context, so the inner `"` does not terminate the outer string — POSIX-specified, not incidental.

```
$ bash -n t.sh   -> parses cleanly
$ bash -e t.sh   -> exit 0
$ cat out.txt
next=1.4.0
bump=minor
current=1.3.0
```

So the step would not have failed at runtime and no release was at risk.

I rewrote it anyway, as a single `node -e` emitting both lines — not because it was broken, but because two readers independently read it as broken, and shell nobody trusts is shell someone "fixes" incorrectly later. Flagging the reasoning so this does not read as agreement.

File: .github/workflows/release.yml [hunk 5]
```diff
@@ -1,29 +1,30 @@
...
-  # Lets the publish path be validated without minting a tag. A tag is a
-  # permanent, public artifact; discovering a broken pipeline by pushing one and
-  # watching it fail is how v1.3.0 ended up tagged but unpublished.
   workflow_dispatch:
     inputs:
       dry_run:
         description: "Run every check and stop before publishing"
         type: boolean
-        default: true
+        default: false
```
[the remark this answers is not in the document]
Comment (marius-cetanas) [reply .github/workflows/release.yml]: Agreed, and this one is the most consequential finding in the review — I had missed it entirely, because I was reading review bodies and this arrived inline.

Defaulting `dry_run` to `false` meant that clicking "Run workflow" without reading the form performed a real publish. npm will not let a version be reissued, so that is unrecoverable rather than merely inconvenient, and it made the safe path the one requiring attention. `release-prepare` already defaulted to a dry run, so the two halves of the process disagreed about which way was safe.

Now defaults to `true`, with the description spelling out that unchecking it is what actually ships. A test asserts the default, since this is the kind of thing that gets flipped later for convenience.

The friction this adds is a feature: a dry run first means the checks — preflight, version-not-already-published, build, coverage, audit — have all passed before anyone commits to an irreversible action.

**Comment by marius-cetanas** [comment]
Round 9 — one finding, plus an inline one I had missed. Both fixed in b6277b2.

**`dry_run` defaulted to `false` on the publish workflow** (inline, and the more serious of the two). Clicking "Run workflow" without reading the form would have published for real, and npm will not let a version be reissued — unrecoverable, not merely inconvenient. `release-prepare` already defaulted to a dry run, so the two halves of the process disagreed about which direction was safe. Now defaults to `true`.

My process failure, worth naming: I was reading review bodies each round and missed this because it arrived as an inline comment. Both previously unanswered inline threads now have replies.

**The changelog prepend used `indexOf("\\n## ")` as a slice index without checking for `-1`.** On a changelog with no release headings yet — a first release against a bare Keep a Changelog preamble — that silently dropped the final character and put the entry in the wrong place. Silent corruption of a file nobody re-reads before merging.

Moved to `scripts/changelog-entry.mjs`, nine tests including the preamble-only file, an entirely empty one, and a `## [x]` appearing inside prose rather than as a heading. Verified against a real preamble-only file: preamble intact, entry correctly placed.

**A pattern worth stating.** Every finding against this pipeline has been in shell or node embedded in YAML — the head_commit gate, the awk regex, the exit-code collapse, this. None of it was testable, which is why reading was the only thing catching it. This is the third such block extracted into a script with tests. The workflows are now thin: they sequence steps and call tested code.

391 tests, 100% coverage of `src/` held.

**Comment by marius-cetanas** [comment]
Round 10 — one finding, correct, fixed in the push above.

**A first-line heading was invisible to the search.** `indexOf("\n## ")` cannot match a heading at position 0, so a changelog with no preamble fell through to the `-1` branch and appended the entry at the end — the opposite of the documented behaviour, and silent.

This is the same bug I *introduced* while fixing the `-1` case last round: I handled "no headings at all" and treated that as the only reason the search could fail. It has two causes, and I only fixed one.

Now finds the first line that **is** a heading rather than searching for a substring. That sees a first-line heading, still ignores `## [x]` appearing inside prose, and preserves the append behaviour when there are genuinely no release headings.

Verified both shapes end to end rather than only in unit tests:

- no-preamble file → new entry first, previous release intact below
- this repo's CHANGELOG → preamble intact, entry above the newest release

Three more tests for the no-preamble shape, twelve on the extractor overall. 394 tests, 100% coverage of `src/` held, no unresolved review threads.

**Comment by marius-cetanas** [comment]
Round 11 — two findings, both correct, fixed in d4a0fa8.

**The `_authToken` pattern matched mid-line.** Right, and `cafile=/tmp/:_authToken=oops` defeats it exactly as described.

This is the third round on this one regex, and the pattern of my failures is the interesting part. It went `(^|[:/])` → `(^|:)` → `(^|/:)`, each round narrowing the *prefix* — and every version still matched anywhere in the line, because the prefix was never the problem. An `.npmrc` key is anchored at the start of the line. Anything to the right of a `=` is a value and cannot be an assignment, however it is punctuated. It is now `^(\/\/.*\/:)?_authtoken\s*=`: a bare key, or a registry-scoped one where the whole key still begins the line.

I was fixing the symptom each round rather than the shape, and the tests I added each round encoded that same mistake — which is why they kept passing. The new tests are table-driven over both directions, and deliberately include the four values that defeated earlier revisions:

```
does not match:  cafile=/tmp/:_authToken=oops
                 cafile=C:_authToken=oops
                 cafile=/etc/ssl/_authToken=oops
                 prefix=//host/:_authToken=oops
                 //host/:_authToken_backup=x

still matches:   _authToken=x
                 //registry.npmjs.org/:_authToken=x
                 //npm.pkg.github.com/:_authToken=x
                 //registry.npmjs.org/some/path/:_authToken=
```

Verified end to end too: the false-positive file exits 0, a real assignment exits 1.

**`execFileSync` status can be null.** Correct — null when the child dies from a signal, which breaks the `code: number` contract the helper declares and would make `toContain(code)` behave oddly. Coerced to 1 for non-numeric statuses.

403 tests, 100% coverage of `src/` held.

**Comment by marius-cetanas** [comment]
Round 12 — one finding, correct, fixed in the push above.

**A missing `--current` threw a stack trace instead of printing usage.** Both other script CLIs in this pipeline handle it, so this was inconsistent as well as unhelpful — and a stack trace is a poor first contact with a tool someone is running by hand precisely because something has already gone wrong.

Now prints usage and exits 1, which is deliberately distinct from the 2 that means "nothing releasable" — the workflow branches on that difference, so collapsing them would reintroduce the misdiagnosis fixed in round 7.

All four paths verified as distinct:

```
missing --current -> exit 1  + usage on stderr
malformed version -> exit 1
releasable        -> exit 0
nothing to ship   -> exit 2
```

Two tests: usage is printed with no stack frame in stderr, and the code is 1 rather than 2. The helper now captures stderr, which it had been discarding — which is why nothing caught this.

405 tests, 100% coverage of `src/` held, no unresolved threads.
